content/blog/2023-04-04-the-pci-device-that-wasnt.md

```diff
@@ -0,0 +1,289 @@
+
+I'd never heard of the [VGA](https://en.wikipedia.org/wiki/VGA_connector) arbiter in the kernel (do kids these days know what VGA is?), or `vgaarb` as it's called.  What it does is irrelevant to this bug, but I found the history pretty interesting!  [Benjamin Herrenschmidt proposed VGA arbitration back in 2005](https://lists.freedesktop.org/archives/xorg/2005-March/006663.html) as a way of resolving conflicts between multiple legacy VGA devices that want to use the same address assignments.  This was previously handled in userspace by the X server, but issues arose with multiple X servers on the same machine.  Plus, it's probably not a good idea for this kind of thing to be handled by userspace.  [You can read more about the VGA arbiter in the kernel docs](https://docs.kernel.org/gpu/vgaarbiter.html), but it's probably not something anyone has thought much about in a long time.
+
+
+# The bad access
+
+```c
+static bool vga_arbiter_add_pci_device(struct pci_dev *pdev)
+{
+        struct vga_device *vgadev;
+        unsigned long flags;
+        struct pci_bus *bus;
+        struct pci_dev *bridge;
+        u16 cmd;
+
+        /* Only deal with VGA class devices */
+        if ((pdev->class >> 8) != PCI_CLASS_DISPLAY_VGA)
+                return false;
+```
+
+We're blowing up on the read to `pdev->class`, and it's not something like the data being uninitialised, it's out-of-bounds.  If we look back at the call trace:
+
+```text
+vga_arbiter_add_pci_device+0x60/0xe00
+pci_notify+0x88/0x444
+notifier_call_chain+0x104/0x320
+blocking_notifier_call_chain+0xa0/0x140
+device_add+0xac8/0x1d30
+device_register+0x58/0x80
+vio_register_device_node+0x9ac/0xce0
+vio_bus_scan_register_devices+0xc4/0x13c
+```
+
+This thing is a VIO device, not a PCI device!  Let's jump into the caller, `pci_notify()`, to find out how we got our `pdev`.
+
+```c
+static int pci_notify(struct notifier_block *nb, unsigned long action,
+                      void *data)
+{
+        struct device *dev = data;
+        struct pci_dev *pdev = to_pci_dev(dev);
+```
+
+So `pci_notify()` gets called with our VIO device (somehow), and we're converting that `struct device` into a `struct pci_dev` with no error checking.  We could solve this particular bug by just checking that our device is *actually* a PCI device before we proceed - but we're in a function called `pci_notify`, we're expecting a PCI device to come in, so this would just be a bandaid.
+
+`to_pci_dev()` works like other struct containers in the kernel - `struct pci_dev` contains a `struct device` as a member, so the `container_of()` function returns an address based on where a `struct pci_dev` would have to be if the given `struct device` was actually a PCI device.  Since we know it's not actually a PCI device and this `struct device` does not actually sit inside a `struct pci_dev`, our `pdev` is now pointing to some random place in memory, hence our access to a member like `class` is caught by KASAN.
+
```

Comment:
Might be too much detail, but it's not a random place in memory, it's somewhere a few 10s-100s of bytes below the `struct device`. And in crucially the location of the `dev` in `struct vio_dev` is much closer to the start of the struct. If the two `foo_dev` structs were the same size KASAN wouldn't have caught it.

Reply:
"Random" definitely wasn't the correct word, "irrelevant" perhaps.  The location of `dev` in the structs is a good point I hadn't considered so I'll write something about that, thanks.